feat(floating-ui): add feature flag for dynamic floating-ui styles

annawen1 opened the pull request:
Closes #17018 

This PR introduces a new feature flag (`enable-v12-dynamic-floating-styles`) to separate the `@floating-ui` logic. By turning on the flag, components will have dynamically positioned floating styles without the collision detection currently in place. 

To have the collision detect logic, adopter will set the `autoAlign` prop to true which will automatically include the dynamic styling - so the feature flag is not necessary in this case.

#### Changelog

**New**

- `enable-v12-dynamic-floating-styles` feature flag

**Changed**

- add checks in current components utilitizing `@floating-ui` to separate out the style and middleware logic

#### Testing / Reviewing

- go to deploy preview, `Feature flags` section
- go to `Dynamic floating styles` folder and inspect the components and see that the styles are set dynamically

annawen1 committed:
feat(feature-flag): create floating styles feature flag

annawen1 committed:
feat(dropdown): add feature flag logic to dropdown and popover

annawen1 committed:
Merge branch 'main' into feat/feature-flag-floating-styles

annawen1 committed:
feat(toggletip): add feature flag story

annawen1 committed:
feat(floating-ui): add additional components

annawen1 committed:
Merge remote-tracking branch 'upstream/main' into feat/feature-flag-floating-styles

annawen1 committed:
chore: shorten flag name

annawen1 committed:
chore: shorten flag name

annawen1 committed:
chore: the rest of the auto align components

annawen1 committed:
chore: set a few controls

annawen1 committed:
Merge branch 'main' into feat/feature-flag-floating-styles

annawen1 committed:
chore: resolve merge conflict

kennylam commented on packages/react/src/components/FeatureFlags/overview.mdx:
```diff
@@ -117,11 +118,14 @@ Feature flags can also be enabled via the provided `enable()` mixin
-
-The `flags` prop will be removed in a future release. Instead, use individual boolean props for each feature flag. 
-A `featureflag-deprecate-flags-prop` codemod has been provided to help deprecate the `flags` object prop and switch to individual boolean props.
+The `FeatureFlags` component has been updated to improve compatibility. The
+`flags` object prop is now deprecated and is replaced with individual boolean
+props for each feature flag.
 
+The `flags` prop will be removed in a future release. Instead, use individual
+boolean props for each feature flag. A `featureflag-deprecate-flags-prop`
+codemod has been provided to help deprecate the `flags` object prop and switch
+to individual boolean props.
```
```suggestion
The `FeatureFlags` component has been updated to improve compatibility. The 
`flags` object prop is now deprecated and will be removed in a future release.

Use individual boolean props for each feature flag instead. A 
`featureflag-deprecate-flags-prop` codemod has been provided to help 
deprecate the `flags` object prop and switch to individual boolean props.
```

This paragraph is kinda clunky and repetitive. What do you think?

annawen1 committed:
Update packages/react/src/components/FeatureFlags/overview.mdx

Co-authored-by: kennylam <909118+kennylam@users.noreply.github.com>

annawen1 committed:
Merge branch 'main' into feat/feature-flag-floating-styles

guidari committed:
Added tests for FluidTimePicker (#17541)

* test: added tests for FluiTimePicker

* fix: changed from fireevent to userevent

riddhybansal committed:
Added typescript types to fluid combobox and its skeleton state (#17563)

* fix: renamed js files to tsx for fluid combobox components

* fix: added typescript types to fluid combobox

* fix: snapshots

* Update packages/react/src/components/FluidComboBox/FluidComboBox.tsx

---------

Co-authored-by: Gururaj J <89023023+Gururajj77@users.noreply.github.com>
Co-authored-by: Taylor Jones <tay1orjones@users.noreply.github.com>
Co-authored-by: Alison Joseph <alison.joseph@us.ibm.com>

annawen1 committed:
chore(storybook): add meta isTemplate to overview doc

dependabot[bot] committed:
chore(deps): bump rollup (#17631)

Bumps [rollup](https://github.com/rollup/rollup) from 4.14.0 to 4.24.0.
- [Release notes](https://github.com/rollup/rollup/releases)
- [Changelog](https://github.com/rollup/rollup/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rollup/rollup/compare/v4.14.0...v4.24.0)

---
updated-dependencies:
- dependency-name: rollup
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alisonjoseph committed:
test(togglesmallskeleton): add test coverage (#17674)

* test(togglesmallskeleton): add test coverage

* chore: fix copyright date

* chore: fix date

emyarod committed:
fix(FileUploaderDropContainer): resolve click handler errorerror (#17568)

Co-authored-by: Alison Joseph <alison.joseph@us.ibm.com>

Gururajj77 committed:
feat: combobox's new feature, autocomplete with typeahead (#17268)

* feat: autocomplete is fked

* feat: autoComplete typeahead feature fixed

* feat: added tests for autocomplete

* feat: adds functionality test cases

* feat: matcsh case exactly with option list when Tab is pressed

* feat: adds tests, refines functions

* feat: updated prop name to component and tests

* refactor: fixed conflict

* test: covered allowcustomValue prop

---------

Co-authored-by: Nikhil Tomar <63502271+2nikhiltom@users.noreply.github.com>
Co-authored-by: Nikhil Tomar <nikhiltomar753@gmail.com>

annawen1 committed:
chore(storybook): move feature flag overview under welcome (#17685)

abdonrd committed:
docs(web-components): remove old storybook copy (#17655)

janhassel committed:
docs(date-time-format): update link to carbon-for-products page (#17692)

emyarod reviewed (approved):
looks good to me!

ariellalgilmore committed:
test(tabs): increase coverage (#17565)

ariellalgilmore committed:
fix(date-picker): clearing input and selecting a date (#17669)

* fix(date-picker): keep open when clearing input

* fix(date-picker): add type single to condition

ariellalgilmore committed:
chore(chat-button-toggle-skeleton): test updates (#17440)

carbon-automation[bot] committed:
chore(release): v11.68.0 (#17696)

Co-authored-by: alisonjoseph <2753488+alisonjoseph@users.noreply.github.com>

2nikhiltom committed:
test: increase test coverage to 100% for Tag (#17614)

* test: increase test coverage to 100% for Tag

* refactor: spyOn updated

* refactor: removed unused import

---------

Co-authored-by: Preeti Bansal <146315451+preetibansalui@users.noreply.github.com>

2nikhiltom committed:
fix: increase test coverage for ToggleTip (#17661)

* fix: increase test coverage for ToggleTip

* refactor: userEvent over fireEvent

---------

Co-authored-by: Alison Joseph <alison.joseph@us.ibm.com>

tay1orjones committed:
chore(issue-templates): always apply needs triage label (#17681)

* chore(issue-templates): always apply needs triage label

* chore: add web components package to package selection dropdown

preetibansalui committed:
fix: link overview page is not loading (#17710)

ariellalgilmore committed:
test(structured-list): increase coverage (#17705)

alisonjoseph committed:
fix(tabs): add icon only 1px spacing (#17671)

* fix(tabs): add icon only 1px spacing

* chore: cleanup story updates

tay1orjones committed:
ci(actions): add workflow to calculate and surface PR merge rate (#17684)

* ci(actions): add workflow to calculate and surface PR merge rate

* ci(actions): add workflow to calculate and surface PR merge rate

---------

Co-authored-by: Nikhil Tomar <63502271+2nikhiltom@users.noreply.github.com>
Co-authored-by: Alison Joseph <alison.joseph@us.ibm.com>

emyarod committed:
fix(TreeNode): preserve nested node expand state (#17630)

* fix(TreeNode): preserve nested node expand state

* fix(TreeView):  prevent focus on hidden child nodes

soumyaraju committed:
feat(tabs): added hover and active styles (#16523)

* feat(tabs): added hover and active styles

* feat(tabs): revert test style

* feat(tabs): active tab style order fix

* feat(tabs): issue fix :last tab is being partially hidden

* feat(tabs): issue fix :last tab is being partially hidden

* feat(tabs): added to contributors list

* feat(tabs): rebase

---------

Co-authored-by: Soumya Raju <soumyaraju@ip-192-168-1-8.us-west-2.compute.internal>
Co-authored-by: Alison Joseph <alison.joseph@us.ibm.com>

renovate[bot] committed:
chore(deps): update dependency @types/prop-types to v15.7.13 (#17657)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Guilherme Datilio Ribeiro <guilhermedatilio@gmail.com>

ariellalgilmore committed:
fix(combobox): rendering aria-label (#17678)

* fix(combobox): rendering aria-label

* test(combobox): update tests with aria-label and titleText

tay1orjones commented on packages/react/src/components/Popover/DynamicStyles.featureflag.mdx:
```diff
@@ -0,0 +1,38 @@
+is needed but the collision detection is not desired.
+
+**Note**: If the `autoAlign` prop is used, the feature flag won't be necessary
+as the collision detection includes the dynamic styles logic.
```
```suggestion
Components with `autoAlign` use the
[`floating-ui`](https://github.com/floating-ui/floating-ui) library to handle
the dynamic positioning of floating elements relative to their trigger / anchor
element. This also includes collision detection that repositions the floating
element as necessary to keep it visible within the viewport. In the majority of
cases, the styling from the `floating-ui` 'fixed' positioning strategy prevents
the floating element from being
[clipped](https://floating-ui.com/docs/misc#clipping) by an ancestor element.

The `enable-v12-dynamic-floating-styles` flag enables this dynamic styling
without the collision detection.

**Note**: The flag has no effect when used with `autoAlign` because `autoAlign`
includes both the dynamic positioning styles and collision detection.
```

tay1orjones reviewed (approved):
LGTM with one minor suggestion

Shankar-CodeJunkie committed:
fix: implement unique dynamic id generation for FileUploader component (#17651)

* fix: implement unique dynamic id generation for FileUploader component

* test(FileUploader): fix failing tests and update snapshot

annawen1 committed:
Update packages/react/src/components/Popover/DynamicStyles.featureflag.mdx

Co-authored-by: Taylor Jones <tay1orjones@users.noreply.github.com>
